src/psset.c

```diff
@@ -0,0 +1,211 @@
+	    /* slab */ false, SC_NSIZES, /* sn */ 0, extent_state_active,
+	    /* zeroed */ false, /* committed */ true, EXTENT_PAI_HPA,
+	    EXTENT_NOT_HEAD);
+	edata_ps_set(r_edata, ps);
+	fb_set_range(ps_fb, ps_npages, begin, npages);
+	/*
+	 * OK, we've got to put the pageslab back.  First we have to figure out
+	 * where, though; we've only checked run sizes before the pageslab we
+	 * picked.  We also need to look for ones after the one we picked.  Note
+	 * that we want begin + npages as the start position, not begin + len;
+	 * we might not have used the whole range.
+	 */
+	start = begin + npages;
+	while (start < ps_npages) {
+		bool found = fb_urange_iter(ps_fb, ps_npages, start, &begin,
+		    &len);
+		if (!found) {
+			break;
+		}
+		if (len > largest_unchosen_range) {
+			largest_unchosen_range = len;
+		}
+		start = begin + len;
+	}
+	edata_longest_free_range_set(ps, (uint32_t)largest_unchosen_range);
+	if (largest_unchosen_range != 0) {
+		psset_insert(psset, ps, largest_unchosen_range);
+	}
+}
+
+bool
+psset_alloc_reuse(psset_t *psset, edata_t *r_edata, size_t size) {
+	edata_t *ps = psset_recycle_extract(psset, size);
+	if (ps == NULL) {
+		return true;
+	}
+	psset_ps_alloc_insert(psset, ps, r_edata, size);
+	return false;
+}
+
+void
+psset_alloc_new(psset_t *psset, edata_t *ps, edata_t *r_edata, size_t size) {
+	fb_group_t *ps_fb = edata_slab_data_get(ps)->bitmap;
+	size_t ps_npages = edata_size_get(ps) >> LG_PAGE;
+	assert(fb_empty(ps_fb, ps_npages));
+
+	assert(ps_npages >= (size >> LG_PAGE));
+	psset_ps_alloc_insert(psset, ps, r_edata, size);
+}
+
+edata_t *
+psset_dalloc(psset_t *psset, edata_t *edata) {
+	assert(edata_pai_get(edata) == EXTENT_PAI_HPA);
+	assert(edata_ps_get(edata) != NULL);
+
+	edata_t *ps = edata_ps_get(edata);
+	fb_group_t *ps_fb = edata_slab_data_get(ps)->bitmap;
+	size_t ps_old_longest_free_range = edata_longest_free_range_get(ps);
+
+	size_t ps_npages = edata_size_get(ps) >> LG_PAGE;
+	size_t begin =
+	    ((uintptr_t)edata_base_get(edata) - (uintptr_t)edata_base_get(ps))
+	    >> LG_PAGE;
+	size_t len = edata_size_get(edata) >> LG_PAGE;
+	fb_unset_range(ps_fb, ps_npages, begin, len);
+
+	/* We might have just created a new, larger range. */
+	size_t new_begin = (size_t)(fb_fls(ps_fb, ps_npages, begin) + 1);
+	size_t new_end = fb_ffs(ps_fb, ps_npages, begin + len - 1);
+	size_t new_range_len = new_end - new_begin;
+	/*
+	 * If the new free range is no longer than the previous longest one,
+	 * then the pageslab is non-empty and doesn't need to change bins.
+	 * We're done, and don't need to return a pageslab to evict.
+	 */
+	if (new_range_len <= ps_old_longest_free_range) {
+		return NULL;
+	}
+	/*
+	 * Otherwise, it might need to get evicted from the set, or change its
+	 * bin.
+	 */
+	edata_longest_free_range_set(ps, (uint32_t)new_range_len);
+	/*
+	 * If it was previously non-full, then it's in some (possibly now
+	 * incorrect) bin already; remove it.
```

Comment:
It looks we may remove and insert into the same bin, I feel it might not happen that rare and we should try optimizing for the case (in a later PR would be fine too)

Reply:
That's a good point. We kinda sorta do that above, with the `new_range_len <= ps_old_longest_free_range` check; but it's possible that we did expand the largest range, but not by enough to bump it into the next size class.